File: Python/0094-binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Iterative
# Time O(n)
# Space O(h)
class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        result = []
        stack = []
        cur = root
        while stack or cur:
            if cur:
                stack.append(cur)
                cur = cur.left
            else:
                cur = stack.pop()
                if not cur:
                    result.append('null')
                else:
                    result.append(cur.val)
                cur = cur.right
        return result



# Recursive
# Time O(n), T(n) = 2*T(n/2)+1
# Space O(h), worst case O(n), average case O(logn)
class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        def helper(root):
            nonlocal res
            if not root:
                return

            helper(root.left)
            res.append(root.val)
            helper(root.right)

        if not root:
            return []
        res = []
        helper(root)
        return res

File: C++/0094-binary-tree-inorder-traversal.cpp
// Time: O(n)
// Space: O(1)
// Morris inorder traversal
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> nums;
        TreeNode *cur = root;
        while (cur) {
            if (cur->left) {
                // find predecessor
                TreeNode *pred = cur->left;
                // keep going right till right node is not null or 
                // right node is not current.
                while (pred->right != cur && pred->right)
                    pred = pred->right;
                if (!pred->right) {
                    // (forward)
                    // go left after establishing link from predecessor
                    // to current.
                    pred->right = cur;
                    cur = cur->left;
                } else {
                    // (backtrack)
                    // left is already visit. Go right after visiting current.
                    pred->right = nullptr;
                    nums.push_back(cur->val);
                    cur = cur->right;
                }
            } else {
                // visit and go to right
                nums.push_back(cur->val);
                cur = cur->right;
            }
        }
        return nums;
    }
};
